inc/vcf/record_cache.hpp

```diff
@@ -0,0 +1,127 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef VCF_VALIDATOR_RECORDCACHE_HPP
+#define VCF_VALIDATOR_RECORDCACHE_HPP
+
+
+#include <set>
+#include <sstream>
+#include "normalizer.hpp"
+#include "file_structure.hpp"
+
+namespace ebi
+{
+  namespace vcf
+  {
+
+    /**
+     * Stores a summary of a Record to check that there are no duplicates.
+     *
+     * For memory reasons, this class can be configured to store only the last `n` elements. This will only detect
+     * duplicates if the input is almost sorted (i.e. if no element is unsorted out of its place more than `n` elements)
+     */
+    class RecordCache
+    {
+      public:
+        RecordCache() : RecordCache{1000} { }
+
+        /**
+         * @param capacity: maximum amount of RecordCores that this instance can hold at any time.
+         * Default is 1000.
+         * A value of 0 is valid but is useless.
+         * A negative value disables the limit, thus storing every RecordCore received, use with caution.
+         */
+        RecordCache(long capacity)
+        {
+            if (capacity >= 0) {
+                this->capacity = static_cast<size_t>(capacity);
+                unlimited = false;
+            } else {
+                this->capacity = 0;
+                unlimited = true;
+            }
+        }
+
+        /**
+         * returns a vector of RecordCores that are duplicated within the previously passed Records.
+         *
+         * See the constructors for the capacity description.
+         * The goal is to report all occurrences of a duplicate record just once each.
+         * if one record already processed is equivalent to the new one, both that and the new record are reported in 2 Errors.
+         * if more than one record already processed is equivalent to the new one, only the new one is reported.
+         * Nonetheless, if the capacity is too small, it may cause incorrect reporting, such as reporting several times
+         * the first occurrence or failing to report duplicates that are farther apart than the capacity.
+         */
+        std::vector<std::unique_ptr<Error>> check_duplicates(const Record &record)
+        {
+            auto record_cores = normalize(record);
+            std::vector<std::unique_ptr<Error>> duplicates{};
+
+            for (RecordCore &record_core: record_cores) {
+                std::pair<std::multiset<RecordCore>::iterator, std::multiset<RecordCore>::iterator> range =
+                        cache.equal_range(record_core);
+
+                if (range.first == range.second) {
+                    // no matches found
+                } else {
+                    // one or more matches found
+                    std::stringstream ss;
+                    ss << "Duplicated record detected: first occurrence: " << *range.first
+                    << ", second occurrence: " << record_core;
```

Comment:
The way the fields are displayed is not very clear for an end-user: the line number and the variant fields should be split instead of displayed all together between brackets. This message could be simplified to something like:

> "Duplicated variant " << record_core << " found in lines " << *range.first.line << " and " << record_core.line